### PR TITLE
Bug 1855341: ovirt: Validate api and apps FQDN before cluster provisioning

### PR DESCRIPTION
**Pull request by eslutsky**
perform DNS resolution  validation to  both the API and the apps fqdn .

Bug-Url: https://bugzilla.redhat.com/show_bug.cgi?id=1855341

Signed-off-by: Evgeny Slutsky <eslutsky@redhat.com>

### CODE REVIEW

File: pkg/asset/installconfig/ovirt/validation.go [hunk 1]
```diff
@@ -125,3 +126,22 @@ func validateVNICProfile(platform ovirt.Platform, con *ovirtsdk.Connection) erro
...
 	return nil
 }
+
+// ValidateProvisioning - Perform Provisioning Validations for oVirt
+func ValidateProvisioning(ic *types.InstallConfig) error {
+	clusterName := ic.ObjectMeta.Name
+	zoneName := ic.BaseDomain
+	validFqdns := []string{
+		fmt.Sprintf("api.%s.%s", clusterName, zoneName),
+		fmt.Sprintf("test.apps.%s.%s.com", clusterName, zoneName),
```
Comment (rgolangh) [inline pkg/asset/installconfig/ovirt/validation.go:136]: .com?

Comment (eslutsky) [reply pkg/asset/installconfig/ovirt/validation.go]: yeap sorry , i've added for my local testing and didn't remove

File: pkg/asset/installconfig/ovirt/validation.go [hunk 2]
```diff
@@ -125,3 +126,22 @@ func validateVNICProfile(platform ovirt.Platform, con *ovirtsdk.Connection) erro
 	}
 	return nil
 }
+
+// ValidateProvisioning - Perform Provisioning Validations for oVirt
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/ovirt/validation.go:130]: this does not follow the guidelines for go function comments. https://blog.golang.org/godoc

File: pkg/asset/installconfig/ovirt/validation.go [hunk 3]
```diff
@@ -125,3 +126,22 @@ func validateVNICProfile(platform ovirt.Platform, con *ovirtsdk.Connection) erro
...
+func ValidateProvisioning(ic *types.InstallConfig) error {
+	clusterName := ic.ObjectMeta.Name
+	zoneName := ic.BaseDomain
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/ovirt/validation.go:133]: there is already a cluster domain function on the install config for this AFAIR

File: pkg/asset/installconfig/ovirt/validation.go [hunk 4]
```diff
@@ -125,3 +126,22 @@ func validateVNICProfile(platform ovirt.Platform, con *ovirtsdk.Connection) erro
...
+		fmt.Sprintf("api.%s.%s", clusterName, zoneName),
+		fmt.Sprintf("test.apps.%s.%s", clusterName, zoneName),
+	}
+
+	for _, addr := range validFqdns {
+		_, err := net.LookupIP(addr)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/ovirt/validation.go:144]: I think we should be returning better error messages.

File: pkg/asset/installconfig/platformprovisioncheck.go [hunk 5]
```diff
@@ -70,8 +71,15 @@ func (a *PlatformProvisionCheck) Generate(dependencies asset.Parents) error {
...
-	case aws.Name, libvirt.Name, none.Name, openstack.Name, ovirt.Name:
+	case aws.Name, libvirt.Name, none.Name, openstack.Name:
 		// no special provisioning requirements to check
+
+	case ovirt.Name:
+		err = icovirt.ValidateProvisioning(ic.Config)
+		if err != nil {
+			return err
+		}
+
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/platformprovisioncheck.go:82]: move this up, lets leave the no-work and default at the end.

Comment (eslutsky) [reply pkg/asset/installconfig/platformprovisioncheck.go]: done

File: pkg/asset/installconfig/ovirt/validation.go (hunk 4 above)
Comment (eslutsky) [reply pkg/asset/installconfig/ovirt/validation.go]: done

File: pkg/asset/installconfig/ovirt/validation.go (hunk 2 above)
Comment (eslutsky) [reply pkg/asset/installconfig/ovirt/validation.go]: fixed

File: pkg/asset/installconfig/ovirt/validation.go (hunk 3 above)
Comment (eslutsky) [reply pkg/asset/installconfig/ovirt/validation.go:133]: not sure what you mean.. this function invoked on  ovirt validation only

**Comment by sandrobonazzola** [comment]
@eslutsky Please fix PR title with prefix: "Bug 1855341: "

**Review by sandrobonazzola — APPROVED** [review]

**Comment by sdodson** [comment]
@abhinavdahiya I know we'd been resistant to adding DNS validation because we didn't assume that the Installer had the same DNS view as the cluster would, are we saying that we have a different opinion now for on-prem clouds or are we saying that we'll validate everywhere? Should this be a warning or an error?

**Comment by abhinavdahiya** [comment]
> @abhinavdahiya I know we'd been resistant to adding DNS validation because we didn't assume that the Installer had the same DNS view as the cluster would, are we saying that we have a different opinion now for on-prem clouds or are we saying that we'll validate everywhere? Should this be a warning or an error?

The installer needs to access the api.clusterdomain to contact the apiserver to check the progress for bootstrapping etc, so it would make sense to say the installer binary host needs access to the DNS. For our om-prem environements there seems to a required at least for existence of the api.clusterdomain DNS name pointing to the api vip afaik so there should be no harm in validating that before we start creating the cluster.
I think the warning should be acceptable too.

we might we better served to remove that requirement imo than to add this validation as the error is pretty apparent in the error after the resources are created and users can run wait-for bootstrap-complete, install-complete to pickup where they left off because of the invalid setup.

**Comment by Gal-Zaidman** [comment]
@eslutsky can you close this one?

**Pull request closed without merging**